Task 03

numatay opened the pull request:
Closes [Task-03](https://github.com/rolling-scopes-school/nodejs-aws-tasks/blob/main/task3-product-api/task.md)

Main tasks:
1 - product-service serverless.yml contains configuration for 2 lambda functions, API is not working at all, but YAML configuration is correct - **DONE**

2 - The getProductsList OR getProductsById lambda function returns a correct response (POINT1) - **DONE**
The URL for getProductsList is https://g7wiw0v6af.execute-api.eu-west-1.amazonaws.com/dev/products

3 - The getProductsById AND getProductsList lambda functions return a correct response code (POINT2) - **DONE**
The URL for getProductById is https://g7wiw0v6af.execute-api.eu-west-1.amazonaws.com/dev/products/5b718fac-8a4c-48b5-9379-215f1526598c

4 - Your own Frontend application is integrated with product service (/products API) and products from product-service are represented on Frontend. AND POINT1 and POINT2 are done. - **DONE**
The link to PR on frontend repo (https://github.com/rolling-scopes-school/nodejs-aws-fe/pull/134)
The deployed version of the web app with backend integration is here: https://d5bzm0spt9d1j.cloudfront.net/


Additional (optional) tasks (but nice to have):
+1 - Async/await is used in lambda functions - **DONE**
+1 - ES6 modules are used for product-service implementation - **DONE**
+1 - Webpack is configured for product-service - **DONE**
+1 - SWAGGER documentation is created for product-service - **NOT_DONE**
+1 - Lambda handlers are covered by basic UNIT tests (NO infrastructure logic is needed to be covered) (You may use JEST) - **NOT_DONE**
+1 - Lambda handlers (getProductsList, getProductsById) code is written not in 1 single module (file) and separated in codebase. - **DONE**
+1 - Main error scenarios are handled by API ("Product not found" error, try-catch blocks are used in lambda handlers). - **DONE**

Commit:
feat: Add 2 methods for products service

Commit:
feat: Add 2 methods for products service

numatay commented:
Dear reviewers,
I haven't pushed the latest version of the task. Please like this post so I can get to once the task will be updated.

Commit:
feat: Add complete implementation of getProductById and getAllProducts services

Commit:
feat: Add CORS headers and update seed

LoraBard commented:
Hi. @numatay seems like this URL is broken: [https://d5bzm0spt9d1j.cloudfront.net/admin/products](https://d5bzm0spt9d1j.cloudfront.net/admin/products). Please go to "Manage products" and check.

numatay commented:
> Hi. @numatay seems like this URL is broken: https://d5bzm0spt9d1j.cloudfront.net/admin/products. Please go to "Manage products" and check.

Yeah, I'm looking at it right now. I have integrated only the front page with aws backend.

LoraBard commented:
> > Hi. @numatay seems like this URL is broken: https://d5bzm0spt9d1j.cloudfront.net/admin/products. Please go to "Manage products" and check.
> 
> Yeah, I'm looking at it right now. I have integrated only the front page with aws backend.

@numatay Just replace /product to /products in this URL [https://ibb.co/BsTkDG4](https://ibb.co/BsTkDG4)

numatay commented:
> > > Hi. @numatay seems like this URL is broken: https://d5bzm0spt9d1j.cloudfront.net/admin/products. Please go to "Manage products" and check.
> > 
> > 
> > Yeah, I'm looking at it right now. I have integrated only the front page with aws backend.
> 
> @numatay Just replace /product to /products in this URL https://ibb.co/BsTkDG4

Yeah, I just fixed the endpoints. Updated the front-end PR and redeployed stack. It works now

LoraBard commented on product-service/src/db.ts:
```diff
@@ -0,0 +1,38 @@
+    products: Map<string, Product>;
+
+    constructor() {
+        this.products = new Map<string, Product>();
```
Possible, but a bit complicated. Did you think about complexity of this solution? Just:
private products = productsList;

getProducts(): Product[] {
return this.products;
}

getProductById(): Product {
const product = this.products.find((item: Product) => item.id === id);
...
}

numatay commented:
I accidentally closed this PR. Unfortunately, once the PR is closed, you cannot re-open it. I decided to create a new PR with the same description so you can leave comments and approve it. Please review [this PR](https://github.com/numatay/shop-backend/pull/2).

LoraBard commented on product-service/src/db.ts:
```diff
@@ -0,0 +1,38 @@
+        });
+    }
+
+    async getProducts() {
```
Why? Async without await? Just make function return mocked data like:
function getMockProducts(): Promise<Product[]> {
return Promise.resolve(productsList);
} and getMockProductsById or util function returns Promise.resolve(data);

LoraBard commented on product-service/src/getProductById.ts:
```diff
@@ -0,0 +1,26 @@
+        const product = await db.getProductById(productId as string);
+        return {
+            statusCode: 200,
+            headers: {
```
It will be better to create handleResponse module for duplicate code

LoraBard commented on product-service/package.json:
```diff
@@ -0,0 +1,30 @@
+  "description": "Serverless webpack example using Typescript",
+  "main": "handler.js",
+  "scripts": {
+    "deploy": "serverless deploy -s dev -r eu-west-1 -v",
```
or you can just move region into provider (serverless.ts)

LoraBard commented on product-service/src/getProductById.ts:
```diff
@@ -0,0 +1,26 @@
+import db from "./db";
+
+export const getProductById = async (event, _context) => {
```
Please add types

LoraBard commented on product-service/src/getAllProducts.ts:
```diff
@@ -0,0 +1,14 @@
+import db from './db';
+import 'source-map-support/register';
+
+export const getAllProducts = async (_event, _context) => {
```
Please add types
